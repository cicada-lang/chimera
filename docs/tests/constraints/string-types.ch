print find _ {
  String("A")
}

print find _ {
  String(1)
}

print find x {
  Equal(x, "cat")
  String(x)
}

// Reification.

print find x {
  String(x)
}

// An unification breaks a type constraint.

print find x {
  String(x)
  Equal(x, 1)
}

// An indirect unification breaks a type constraint.

print find x {
  String(x)
  Equal(x, z)
  Equal(z, 1)
}

print find x {
  Equal(x, z)
  Equal(z, 1)
  String(x)
}

print find x {
  Equal(x, z)
  String(x)
  Equal(z, 1)
}

// An unification subsums a type constraint.

print find x {
  String(x)
  Equal(x, "A")
}

// An indirect unification subsums a type constraint.

print find x {
  String(x)
  Equal(x, z)
  Equal(z, "A")
}

print find x {
  Equal(x, z)
  Equal(z, "A")
  String(x)
}

print find x {
  Equal(x, z)
  String(x)
  Equal(z, "A")
}
