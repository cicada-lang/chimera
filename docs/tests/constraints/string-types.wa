find _ {
  String("A")
}

find _ {
  String(1)
}

find x {
  x = "cat"
  String(x)
}

// Reification.

find x {
  String(x)
}

// An unification breaks a type constraint.

find x {
  String(x)
  x = 1
}

// An indirect unification breaks a type constraint.

find x {
  String(x)
  x = z
  z = 1
}

find x {
  x = z
  z = 1
  String(x)
}

find x {
  x = z
  String(x)
  z = 1
}

// An unification subsums a type constraint.

find x {
  String(x)
  x = "A"
}

// An indirect unification subsums a type constraint.

find x {
  String(x)
  x = z
  z = "A"
}

find x {
  x = z
  z = "A"
  String(x)
}

find x {
  x = z
  String(x)
  z = "A"
}
