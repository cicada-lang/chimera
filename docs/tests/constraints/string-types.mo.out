[_.0]
[]
["cat"]
[_.0 with { $TypeConstraint(String)(_.0) }]
[]
[]
[]
[]
["A"]
["A"]
["A"]
["A"]
